tutorial/src/code/index.js

```diff
@@ -0,0 +1,197 @@
+		plaquetteDx = parseInt(document.getElementById('dxCell').value);
+		plaquetteDy = parseInt(document.getElementById('dyCell').value);
+		libraryTopLeftCorners = [[21, 3], [21, 3+plaquetteDy+2], [21, 3+(plaquetteDy+2)*2], [21, 3+(plaquetteDy*2)*3]]
+		outline.lineStyle(2, 'lightcoral');
+		// Add workspace guidelines.
+		let y0 = guideTopLeftCorner[1];
+		let message = '';
+		while (y0 + plaquetteDy <= GUIDE_MAX_BOTTOM_RIGHT_CORNER_CODE_WORKSPACE[0]) {
+			let x0 = guideTopLeftCorner[0];
+			while (x0 + plaquetteDx <= GUIDE_MAX_BOTTOM_RIGHT_CORNER_CODE_WORKSPACE[1]) {
+				const x1 = x0 + plaquetteDx;
+				const y1 = y0 + plaquetteDy;
+				outline.moveTo(x0*gridSize, y0*gridSize);
+				outline.lineTo(x1*gridSize, y0*gridSize);
+				outline.lineTo(x1*gridSize, y1*gridSize);
+				outline.lineTo(x0*gridSize, y1*gridSize);
+				outline.lineTo(x0*gridSize, y0*gridSize);
+				x0 += plaquetteDx;
+				message += '  .';
+			}
+			y0 += plaquetteDy;
+			message += '\n';
+		}
+		// Add library guidelines.
+		for (const [x0, y0] of libraryTopLeftCorners) {
+			const x1 = x0 + plaquetteDx;
+			const y1 = y0 + plaquetteDy;
+			outline.moveTo(x0*gridSize, y0*gridSize);
+			outline.lineTo(x1*gridSize, y0*gridSize);
+			outline.lineTo(x1*gridSize, y1*gridSize);
+			outline.lineTo(x0*gridSize, y1*gridSize);
+			outline.lineTo(x0*gridSize, y0*gridSize);
+		}
+        // Create the compact representation of the (empty) QEC code
+        const codesummary = document.getElementById('codeSummary');
+        codesummary.value = message;
+		// Add library plaquettes.
+		//const library_workspace = document.getElementsByName('workspace-library');
+		let plaquetteTypes = [];
+		//const numPlaquettes = savedPlaquettes.length;
+		savedPlaquettes.forEach((plaq, index) => {
+			if (plaq.name !== 'WIP plaquette') {
+				let qubits = [];
+				plaq.qubits.forEach((q) => {
+					const qubit = new Qubit(q.globalX, q.globalY, q.Radius);
+					qubit.name = q.name;
+					qubit.updateLabel();
+					workspace.addChild(qubit);
+					qubits.push(qubit);
+				});
+				// Recall that plaquette names are like "plaquette 12", starting from "plaquette 1"
+				const plaquette_id = parseInt(plaq.name.match(/\d+/)[0]);
+				const base_translate_vector = {x: guideTopLeftCorner[0] - libraryTopLeftCorners[plaquette_id-1][0],
+				                               y: guideTopLeftCorner[1] - libraryTopLeftCorners[plaquette_id-1][1]};
+				const p_type = new PlaquetteType(qubits, libraryColors[index], num_background_children, base_translate_vector)
+				p_type.name = plaq.name;
+				plaquetteTypes.push(p_type);
+				workspace.addChildAt(p_type, num_background_children);
+			}
+		});
+	});
+
+/////////////////////////////////////////////////////////////
+
+    // Undo button, meaning that the last plaquette added is removed.
+	const undoButton = button('Remove last plaquette', gridSize, 2*gridSize, 'white', 'black');
+	workspace.addChild(undoButton);
+
+    undoButton.on('click', (_e) => {
```

Comment:
The undo button should track user actions by pushing them onto a stack, and popping them onto the "redo" stack, pushing redo pops the redo stack onto the undo stack, and the top of the redo stack represents the most recent action taken. This allows us to generalize to any user action.

Reply:
This "undo" button is really just for removing the last plaquette (as specified by the button name).
The way I add the plaquette is as children at a specific index using `addChildAt` (I am not sure that index is the right terminology), so removing the child at that index remove the last plaquette (last in, first out).